### PR TITLE
Make `TyS::is_suggestable` check for non-suggestable types structually

### LINKED ISSUE #91832: error help suggests writing down a closure type, which would be a syntax error
Given the following code: [play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn%20f()%20%7B%0A%20%20%20%20vec!%5B%27a%27%5D.iter().map(%7Cc%7C%20c)%0A%7D)

```rust
fn f() {
    vec!['a'].iter().map(|c| c)
}
```

The current output is:

```
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
2 |     vec!['a'].iter().map(|c| c)
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found struct `Map`
  |
  = note: expected unit type `()`
                found struct `Map<std::slice::Iter<'_, char>, [closure@src/lib.rs:2:26: 2:31]>`
help: consider using a semicolon here
  |
2 |     vec!['a'].iter().map(|c| c);
  |                                +
help: try adding a return type
  |
1 | fn f() -> Map<std::slice::Iter<'_, char>, [closure@src/lib.rs:2:26: 2:31]> {
  |        +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

Ideally the output should look like:

Don't suggest adding `-> Map<std::slice::Iter<'_, char>, [closure@src/lib.rs:2:26: 2:31]>` if that's not going to compile.

### PR DESCRIPTION
**Pull request by compiler-errors**
Not sure if I went overboard checking substs in dyn types, etc. Let me know if I should simplify this function.

Fixes #91832

### CODE REVIEW

**Comment by compiler-errors** [comment]
cc: @estebank, who might care about changes in diagnostic type annotations.

File: compiler/rustc_middle/src/ty/diagnostics.rs [hunk 1]
```diff
@@ -63,16 +68,40 @@ impl<'tcx> TyS<'tcx> {
 
     /// Whether the type can be safely suggested during error recovery.
     pub fn is_suggestable(&self) -> bool {
-        !matches!(
-            self.kind(),
+        fn generic_arg_is_suggestible(arg: GenericArg<'_>) -> bool {
+            match arg.unpack() {
+                GenericArgKind::Type(ty) => ty.is_suggestable(),
+                _ => true,
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/diagnostics.rs:74]: it probably makes sense to also return false for `ConstKind::Infer | ConstKind::Bound | ConstKind::Placeholder | ConstKind::Error` here. And also look at array lengths now.

After that r=me unless @estebank or @cjgillot want some other changes here

**Comment by estebank** [comment]
I was sad to see some of the suggestions go away, so I created #91981 that builds on top of this one.

**Pull request merged**